src/poly/commitment.rs

```diff
@@ -48,51 +45,41 @@ impl<C: CurveAffine> Params<C> {
 
         let n: u64 = 1 << k;
 
-        let try_and_increment = |hasher: &Blake2bState| {
-            let mut trial = 0u64;
-            loop {
-                let mut hasher = hasher.clone();
-                hasher.update(&(trial.to_le_bytes())[..]);
-                let mut repr = C::Repr::default();
-                repr.as_mut().copy_from_slice(hasher.finalize().as_bytes());
-                let p = C::from_bytes(&repr);
-                if bool::from(p.is_some()) {
-                    break p.unwrap();
-                }
-                trial += 1;
-            }
-        };
-
-        let g = {
+        let g_projective = {
             let mut g = Vec::with_capacity(n as usize);
-            g.resize(n as usize, C::identity());
+            g.resize(n as usize, C::Curve::identity());
 
             parallelize(&mut g, move |g, start| {
-                let mut hasher = Blake2bParams::new()
-                    .hash_length(32)
-                    .personal(C::BLAKE2B_PERSONALIZATION)
-                    .to_state();
-                hasher.update(b"G vector");
+                let hasher = C::CurveExt::hash_to_curve("Halo2-Parameters");
```

Comment:
It should be possible to create this outside `parallelize`, I think (because it's not mutable).

Reply:
Unfortunately, `parallelize` needs the provided closure to be cloneable (because it spins up multiple threads) and so the hasher needs to impl `Clone` as well. But only `auto` traits like `Sync` and `Send` are allowed to be "additional traits" in trait objects, not traits like `Clone`. We'd have to make a new function trait just for this particular situation, which is annoying.

Reply:
In practice the cost is minor and we can revisit this in the future when GATs are added to Rust or something.